trino/trino_test.go

```diff
@@ -765,6 +765,123 @@ func TestQueryColumns(t *testing.T) {
 	assert.Equal(t, actualTypes, expectedTypes)
 }
 
+func TestMaxTrinoPrecisionDateTime(t *testing.T) {
+	testPrecisionDateTime(t, 12, true)
+}
+
+func TestMaxGoPrecisionDateTime(t *testing.T) {
+	testPrecisionDateTime(t, 9, false)
+}
+
+func testPrecisionDateTime(t *testing.T, precision int64, shouldError bool) {
+
+	c := &Config{
+		ServerURI:         *integrationServerFlag,
+		SessionProperties: map[string]string{"query_priority": "1"},
+	}
+
+	dsn, err := c.FormatDSN()
+	require.NoError(t, err)
+
+	db, err := sql.Open("trino", dsn)
+	require.NoError(t, err)
+
+	t.Cleanup(func() {
+		assert.NoError(t, db.Close())
+	})
+
+	rows, err := db.Query(fmt.Sprintf(`SELECT
+  cast(current_time as time(%d)) AS timep,
+  cast(current_time as time(%d) with time zone) AS timeptz,
+  cast(current_time as timestamp(%d)) AS tsp,
+  cast(current_time as timestamp(%d) with time zone) AS tsptz`, precision, precision, precision, precision))
+	require.NoError(t, err, "Failed executing query")
+	assert.NotNil(t, rows)
+
+	columns, err := rows.Columns()
+	require.NoError(t, err, "Failed reading result columns")
+
+	assert.Equal(t, 4, len(columns), "Expected 4 result column")
+	expectedNames := []string{
+		"timep",
+		"timeptz",
+		"tsp",
+		"tsptz",
+	}
+	assert.Equal(t, expectedNames, columns)
+
+	columnTypes, err := rows.ColumnTypes()
+	require.NoError(t, err, "Failed reading result column types")
+
+	assert.Equal(t, 4, len(columnTypes), "Expected 4 result column type")
+
+	type columnType struct {
+		typeName  string
+		hasScale  bool
+		precision int64
+		scale     int64
+		hasLength bool
+		length    int64
+		scanType  reflect.Type
+	}
+	expectedTypes := []columnType{
+		{
+			"TIME",
+			true,
+			precision,
+			0,
+			false,
+			0,
+			reflect.TypeOf(sql.NullTime{}),
+		},
+		{
+			"TIME WITH TIME ZONE",
+			true,
+			precision,
+			0,
+			false,
+			0,
+			reflect.TypeOf(sql.NullTime{}),
+		},
+		{
+			"TIMESTAMP",
+			true,
+			precision,
+			0,
+			false,
+			0,
+			reflect.TypeOf(sql.NullTime{}),
+		},
+		{
+			"TIMESTAMP WITH TIME ZONE",
+			true,
+			precision,
+			0,
+			false,
+			0,
+			reflect.TypeOf(sql.NullTime{}),
+		},
+	}
+	actualTypes := make([]columnType, 4)
+	for i, column := range columnTypes {
+		actualTypes[i].typeName = column.DatabaseTypeName()
+		actualTypes[i].precision, actualTypes[i].scale, actualTypes[i].hasScale = column.DecimalSize()
+		actualTypes[i].length, actualTypes[i].hasLength = column.Length()
+		actualTypes[i].scanType = column.ScanType()
+	}
+
+	assert.Equal(t, actualTypes, expectedTypes)
+
+	if shouldError {
+		assert.False(t, rows.Next())
+		require.Error(t, rows.Err())
```

Comment:
It would be more readable as two separate tests and you should test for a specific error message, not just the fact that any error is present. If you'd do that, test cases in one of the tests would be simpler as it would not have to check for precision.

Reply:
We can still perfectly get the precision for the columns query even though retrieving the actual rows errors. Also the error message is very technical, and not too helpful because it shows the parsing error for the last layout we tried. Should have a more legible error like "value XXX cannot be parsed as a date or time"? Or collect the various errors into one?

Reply:
> We can still perfectly get the precision for the columns query even though retrieving the actual rows errors.

This is an even better reason to split these up (or just have a separate test for errors).

> Should have a more legible error like "value XXX cannot be parsed as a date or time"

That would be ideal, but if you think it's out of the scope of this PR we can address it later. Is it still possible to test for a specific error message, even if it's a complex one? I mean can we keep the tests simple?

Reply:
Yes, I've changed the test that errors to use a time literal so we can test for a consistent error message since the order of layouts is deterministic. Maybe a better error message could be done in a separate PR, yes.